fix: allow group members to read group information

hugodutka opened the pull request:
Fixes https://github.com/coder/coder/issues/13988. See the issue for background information.

This PR:
- allows Group members to see they are part of the group
- allows Group members to see the total count of that group's members even if they are not permitted to read detailed information about them, like username or email
- exposes a new endpoint under `/organizations/{organization}/users/{user}/reduced-groups` that returns an array of `ReducedGroup` objects. These objects contain basic information about the group and the total member count.
- makes the account page (`/settings/account`) use this endpoint to display information about groups

I asked questions about permission scope in the issue, but didn't get an answer in time, so I assumed that by default:
- group members should be able to see they are part of the group
- they should be able to read the total group members count
- they shouldn't be able to access detailed information about the group members

Let me know if these assumptions are correct.

This is my first PR to Coder, and I'm not that familiar with the codebase, so I'm very much open to feedback and suggestions. :)

hugodutka committed:
- allow group members to read basic Group info
- allow group members to see they are part of the group, but not see that information about other members
- add a GetGroupMembersCountByGroupID SQL query, which allows group members to see members count without revealing other information about the members
- expose a reducedGroupsByUserAndOrganization API endpoint

hugodutka committed:
rename ReducedGroup.member_count to ReducedGroup.total_member_count

hugodutka committed:
use the reduced groups for user endpoint on the AccountPage

hugodutka committed:
rename UserWithGroupAndOrgID to GroupMemberRBACHelper

hugodutka commented:
I have read the CLA Document and I hereby sign the CLA

Emyrk commented on coderd/database/modelmethods.go:
```diff
@@ -173,7 +174,44 @@ func (v TemplateVersion) RBACObjectNoTemplate() rbac.Object {
+	OrganizationID uuid.UUID
+}
+
+func (gm GroupMemberRBACHelper) RBACObject() rbac.Object {
+	return rbac.ResourceGroup.WithID(gm.GroupID).InOrg(gm.OrganizationID).
+		// Group member can see they are in the group.
+		WithACLUserList(map[string][]policy.Action{
+			gm.User.ID.String(): {
+				policy.ActionRead,
+			},
+		})
+}
```
This is a nifty trick. We can accomplish this same thing without using an ACL though if we just make a new resource for group members. 

What we can do is add a new `group_member` resource to [`policy.go`](https://github.com/coder/coder/blob/main/coderd/rbac/policy/policy.go#L144-L151). 

Then we can use `rbac.ResourceGroupMember.WithID(r.GroupID).InOrg(r.OrganizationID).WithOwner(r.UserID)` that will grant an allow because the `WithOwner()` is essentially the same thing as the User ACL trick you did.

There is going to be some ripple effects of the new resource, such as adding the new resource permission to [`userAdminRole`](https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L342-L342), [`orgUserAdmin`](https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L448-L468), and [`orgAdmin`](https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L387-L405).

----

I think you still need to do the `fetch` business here: https://github.com/coder/coder/blob/main/coderd/database/dbauthz/dbauthz.go#L1393-L1407 because `organization_id` is not present in the returned `database.User`.

hugodutka replied:
### Regarding WithOwner

> WithOwner() is essentially the same thing as the User ACL trick you did

I'm having a bit of trouble getting this approach to work. Am I missing something?

The following gives `true` when you run `opa eval --format=pretty -i input.json -d coderd/rbac/policy.rego "data.authz.allow"`

```json
{
    "action": "read",
    "subject": {
        "id": "the_user_id",
        "groups": [],
        "roles": [],
        "scope": {
            "allow_list": [
                "*"
            ],
            "org": {},
            "site": [
                {
                    "action": "*",
                    "negate": false,
                    "resource_type": "*"
                }
            ],
            "user": []
        }
    },
    "object": {
        "acl_group_list": {},
        "acl_user_list": {
            "the_user_id": ["read"]
        },
        "any_org": false,
        "id": "the_org_id",
        "org_owner": "the_org_id",
        "owner": "",
        "type": "group"
    }
}
```

However, the following gives `undefined`:

```json
{
    "action": "read",
    "subject": {
        "id": "the_user_id",
        "groups": [],
        "roles": [],
        "scope": {
            "allow_list": [
                "*"
            ],
            "org": {},
            "site": [
                {
                    "action": "*",
                    "negate": false,
                    "resource_type": "*"
                }
            ],
            "user": []
        }
    },
    "object": {
        "acl_group_list": {},
        "acl_user_list": {},
        "any_org": false,
        "id": "the_org_id",
        "org_owner": "the_org_id",
        "owner": "the_user_id",
        "type": "group"
    }
}
```

### Regarding the `fetch` business

It seems you were looking at the `main` branch judging by the link. [My changes](https://github.com/coder/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L1385) include the `fetch` logic. Or did you mean something else?

Emyrk replied:
> However, the following gives undefined:

Yes, that will fail (undefined) for 2 reasons. First, you are missing the `member` role which is implied to all Coder users.

https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L272-L296

It is always appended in the SQL here: https://github.com/coder/coder/blob/main/coderd/database/queries/users.sql#L236-L236

And it grants `*` to all resources in the user's level.(except a subset). So you get effectively this role for the groups resource. 

```
      {
        "name":"member",
        "display_name":"Member",
        "site":[],
        "org":{},
        "user":[
          {
            "negate":false,
            "resource_type":"group",
            "action":"*"
          }
        ]
      }
```

But **this will still fail** because a group is also owned by an organizatnion: `"org_owner": "the_org_id"`. The rego policy requires you to be a [member of an organization](https://github.com/coder/coder/blob/main/coderd/rbac/policy.rego#L168-L175) to do anything on org owned objects. This is mainly so if you remove a member from an org, they can no longer access their own workspace in said org.

So you also need to add the `organization_member` role, which is also [always implied](https://github.com/coder/coder/blob/main/coderd/database/queries/users.sql#L246-L246) if they are in the org.

With these changes, the `input.json` becomes something like: _I removed the unnecessary perms_
```json
{
  "action":"read",
  "subject":{
    "id":"the_user_id",
    "groups":[

    ],
    "roles":[
      {
        "name":"member",
        "display_name":"Member",
        "site":[
        ],
        "org":{
          "the_org_id": []
        },
        "user":[
          {
            "negate":false,
            "resource_type":"*",
            "action":"*"
          }
        ]
      }
    ],
    "scope":{
      "allow_list":[
        "*"
      ],
      "org":{

      },
      "site":[
        {
          "action":"*",
          "negate":false,
          "resource_type":"*"
        }
      ],
      "user":[

      ]
    }
  },
  "object":{
    "acl_group_list":{

    },
    "acl_user_list":{

    },
    "any_org":false,
    "id":"the_org_id",
    "org_owner":"the_org_id",
    "owner":"the_user_id",
    "type":"group"
  }
}
```

-------

On the fetch business, :facepalm:. I checked out your branch, but the GitLink extension does not work with multiple git origins I suppose. I updated the link to your fork.

hugodutka replied:
Ah, ok. Thank you for the clarification on RBAC, that's very helpful.

However, I still don't quite understand the comment regarding `fetch` even with the updated link. RBAC filtering that respects org id is done on [this line](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L1408). The returned `[]database.User` is already filtered.

Emyrk replied:
@hugodutka yes. I just mean't, I think you still have to write a custom `fetch` function that returns a custom type with all the fields `user_id`, `organizaton_id`, and `id`. 

Unless the custom `view` we talked about in the other comment joins in this information, and we return a `GroupMember` instead of a `database.User`

hugodutka replied:
That's clear now. Returning `GroupMember`s from the new `view` instead of `User`s is a more elegant approach imo.

Emyrk commented on codersdk/groups.go:
```diff
@@ -35,6 +35,18 @@ type Group struct {
 	Source         GroupSource   `json:"source"`
 }
 
+type ReducedGroup struct {
+	ID             uuid.UUID `json:"id" format:"uuid"`
+	Name           string    `json:"name"`
+	DisplayName    string    `json:"display_name"`
+	OrganizationID uuid.UUID `json:"organization_id" format:"uuid"`
+	// How many users are in this group. Shows the total count,
+	// even if the user is not authorized to read group member details.
+	// May be greater than `len(Group.Members)`.
+	TotalMemberCount int    `json:"total_member_count"`
+	AvatarURL        string `json:"avatar_url"`
+}
+
```
Leaving a larger comment about potentially making this a database model based on a better `group_member` view.

Emyrk commented on enterprise/coderd/coderd.go:
```diff
@@ -287,6 +287,15 @@ func New(ctx context.Context, options *Options) (_ *API, err error) {
 				r.Get("/", api.groupByOrganization)
 			})
 		})
+		r.Route("/organizations/{organization}/users/{user}/reduced-groups", func(r chi.Router) {
+			r.Use(
+				apiKeyMiddleware,
+				api.templateRBACEnabledMW,
+				httpmw.ExtractOrganizationParam(api.Database),
+				httpmw.ExtractUserParam(options.Database),
+			)
+			r.Get("/", api.reducedGroupsByUserAndOrganization)
+		})
```
I feel like this is a good candidate to make a more generalized `/groups` database query, and move this particular route to `GET /users/{user}/groups` or something like that.

The current settings page is using `/organizations/{organization}/groups` for a page that should show all groups. So the frontend is incorrect when multi-org is released ([it is in progress, and I just made an issue for this after seeing this PR!](https://github.com/coder/coder/issues/14213)).

-----

Basically there exists a reason to query groups by `user_id`, and a reason to query them by `organization_id`, and maybe both. We should have a database function that takes various filter arguments, so we can use the same backend call.

Not necessary to do in this PR though, it can be a follow up.

-------

For this PR, at the very least, we should move this to the existing `/groups` endpoint on organizations, and just add a query param to filter by user imo. So like:

`/organizations/{organization}/groups/{groupName}?with_member=<user_id>` or something like that? 


https://github.com/coder/coder/blob/main/enterprise/coderd/groups.go#L348-L352

The returned `codersdk.Group` will just have a potentially empty (or 1 element, yourself) `Members` field. And just add `TotalMemberCount` to the existing `codersdk.Group`.

hugodutka replied:
Once we add the `TotalMemberCount` to `codersdk.Group` this route isn't needed anymore and we can keep on using the existing `/organizations/{organization}/groups` on the account page. To keep this PR focused, I think any new endpoints would better be added in a PR related to https://github.com/coder/coder/issues/14213.

Emyrk replied:
@hugodutka happy to just solve the permissions here, and do endpoints after :+1:

Emyrk reviewed (commented):
This has made me realize that groups and `users` are pretty intertwined, but we should probably have a `GroupMember` resource that helps handle some of these permission edge cases.

> they shouldn't be able to access detailed information about the group members

This is true, and maybe we should just handle this all the way at the database layer. I know originally we returned `database.User` because the table `group_members` has almost no information and is basically useless without more info. So we returned the user.

But we could do something like we do for templates. 

Basically, all template queries use the view `templates_with_names` to join in relevant username, icon, etc information.

This ends up being your `ReducedGroup`, and optionally you could do the `TotalMemberCount` there :man_shrugging:.

SQLc config to make the template view the primary model: https://github.com/coder/coder/blob/main/coderd/database/sqlc.yaml#L77-L78

PR example of updating said view: https://github.com/coder/coder/pull/13858/files


P.S: This is touching many parts of the code with some weak documentation and examples. There is an effort to document some of the RBAC stuff here: https://github.com/coder/coder/pull/14065

This PR is solving some fundamental problems with groups that should be fixed, and really happy to see this PR coming in :+1:

johnstcn commented on enterprise/coderd/groups.go:
```diff
@@ -420,3 +420,41 @@ func (api *API) groups(rw http.ResponseWriter, r *http.Request) {
 	httpapi.Write(ctx, rw, http.StatusOK, resp)
 }
+
+// @Summary Get reduced groups for a user in an organization
```
```suggestion
// @Summary Get reduced groups by user and organization
```

hugodutka replied:
Thanks! The route was deleted, so no changes were necessary.

hugodutka commented:
Thank you for the review @Emyrk. Couple of questions:

## GroupMember resource

> we should probably have a GroupMember resource

I considered this before and found some issues:

- We don't have much control over custom roles in customer deployments. Any role with Group read permissions would need to be updated with the GroupMember permissions to keep on working as before. I'm not sure if "custom roles" is a widely used feature, but customers would probably need to be informed about this change.
- We probably should rework the authz checks on [`InsertGroupMember`](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L2614), [`DeleteGroupMemberFromGroup`](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L976), making this a bigger PR

Do you think we should add the new resource given these problems?

## GroupMember DB View

I like this idea. It would also simplify querying for members of the `Everyone` group.

> This ends up being your `ReducedGroup`, and optionally you could do the `TotalMemberCount` there 🤷‍♂️.

If we add the `TotalMemberCount` field to `codersdk.Group` as you suggested then we don't need the `ReducedGroup` type at all, so I don't think we need the column in the view.

Emyrk commented:
> Thank you for the review @Emyrk. Couple of questions:
> 
> ## GroupMember resource
> > we should probably have a GroupMember resource
> 
> I considered this before and found some issues:
> 
> * We don't have much control over custom roles in customer deployments. Any role with Group read permissions would need to be updated with the GroupMember permissions to keep on working as before. I'm not sure if "custom roles" is a widely used feature, but customers would probably need to be informed about this change.

Custom roles is not yet released, it's in experimental. So there are currently no custom_roles in the wild.

You are correct, adding related resources causes an issue when making roles. As granting 1 half of the relation doesn't really make much sense, and there is no obvious linkage between the two.

But this is a flaw of our current RBAC, and it already exists in a few places. The extra resource is easier to understand then the ACL workaround you have in place imo. 

The good news of the custom_roles though is that we control the interface to making custom roles. Currently, it's basically a 1:1 with permissions, but we could implement some nicer UI and auto create these linkages. So the custom_roles endpoint or frontend could assume `group_member.read` if `group.read` is assigned for example.

So tl;dr, this issue exists for other resources today, and I think we can solve this somewhere else.


> * We probably should rework the authz checks on [`InsertGroupMember`](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L2614), [`DeleteGroupMemberFromGroup`](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L976), making this a bigger PR

Actually I do not think we do. When you insert/delete a member from the group, you are not accessing the group member, you are mutating the `ResourceGroup` still imo. A Group is a set of users, if you are mutating the set, then we should use `ResourceGroup`.

If you are reading/updating an individual member, then it makes sense to use `ResourceGroupMember`.

-------


A final note on all of this. If we model this as a `ResourceGroupMember`, we can actually end up doing the authz check in SQL like we do for [`users`](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L2179-L2186).

As deployments grow, we could easily be fetching a few thousand users, and doing 1k authz checks isn't very efficient (they are in memory and can be ~1ms). 

So if we model things as we have been discussing, it's a fast follow up to optimize into a SQL query later.

hugodutka committed:
add the group_members_expanded db view

hugodutka committed:
add comment on migration

hugodutka committed:
rewrite group member queries to use the group_members_expanded view

hugodutka committed:
add the RBAC ResourceGroupMember and add it to relevant roles

hugodutka committed:
rewrite GetGroupMembersByGroupID permission checks

hugodutka committed:
- make the GroupMember type contain all user fields
- fix type issues coming from replacing User with GroupMember in group member queries

hugodutka committed:
simplify RBAC check on group member count

hugodutka committed:
add the MemberTotalCount field to codersdk.Group

hugodutka committed:
remove the reduced groups endpoint, revert frontend to use the groups by org endpoint

hugodutka committed:
revert "simplify RBAC check on group member count" - new query didn't work on empty groups

hugodutka committed:
display `group.total_member_count` instead of `group.members.length` on the account page

hugodutka committed:
adjust `total_member_count` on `MockGroup`

hugodutka committed:
fixes after rebase

hugodutka committed:
simplify RBAC check on GetGroupMembersCountByGroupID

hugodutka commented:
@Emyrk, I adjusted the PR according to your feedback.

List of changes:

- created a `group_members_expanded` database view
- configured sqlc to use the view as the source of the `GroupMember` object
- rewrote group member queries to use the view
- created the RBAC resource `ResourceGroupMember` and added it to roles which had `Group` permissions
- rewrote dbauthz queries to use the new RBAC object
- refactored code which uses the queries to work with the `GroupMember` return type (instead of `User`)
- added `TotalMemberCount` to `codersdk.Group`
- removed the `/organizations/{organization}/users/{user}/reduced-groups` route
- rolled back frontend changes to use the previous groups by organization endpoint

Let me know if this matches what you had in mind.

Also a question: I'm not familiar with the quotes system, but when I was refactoring SQL queries [`GetQuotaAllowanceForUser`](https://github.com/hugodutka/coder/blob/2f6687a4752ba0eddcc936d3edc4631a950273e0/coderd/database/queries/quotas.sql#L1) stood out to me. It ignores the `Everyone` group. Is this intended?

Emyrk commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -724,17 +724,40 @@ func (q *FakeQuerier) getOrganizationMemberNoLock(orgID uuid.UUID) []database.Or
 
 // getEveryoneGroupMembersNoLock fetches all the users in an organization.
-func (q *FakeQuerier) getEveryoneGroupMembersNoLock(orgID uuid.UUID) []database.User {
+func (q *FakeQuerier) getEveryoneGroupMembersNoLock(orgID uuid.UUID) []database.GroupMember {
```
The `FakeQuerier` has a data field `groupMembers                  []database.GroupMember`. This should now be `groupMembers                  []database.GroupMemberTable`

Doing that will review some queries that need to be updated to include the joined information.

Emyrk reviewed (commented):
This is overall looking good :+1: 

I created a test to mess around with some of this in an easier setting:
https://github.com/coder/coder/pull/14223

If I get a review of it into `main`, it'll be easier to use for your testing. Right now I just copy pasted it in to check out the code paths.

Emyrk commented:
My unit test was merged.

You will have to just fixup these test assertions:
https://github.com/coder/coder/blob/8af8c77e2a9eb1abd20f8070fbf22a04a853f88e/coderd/database/dbauthz/groupsauth_test.go#L114-L130

Just wanted to get an actual breaking test from these changes :+1:

hugodutka committed:
fix tests

hugodutka committed:
resolve lint error

hugodutka committed:
update the groupsauth test to work with new group member permissions

hugodutka committed:
fix a mistake in test

hugodutka commented:
@Emyrk I think I updated all the tests. I rebased onto `main` and `make test`, `make lint` pass locally.

I updated `groupsauth_test.go` to work with the new permission system. I saw that you assumed that if a user didn't have `GroupMember` `read` permission then `GetGroupMembersByGroupID` would return an error, but my implementation returns an empty list instead. I updated the test to match this behavior. Hope that's ok.

Emyrk commented on coderd/database/dbmem/dbmem.go:
```diff
@@ -723,18 +723,68 @@ func (q *FakeQuerier) getOrganizationMemberNoLock(orgID uuid.UUID) []database.Or
 	return members
 }
 
+var ErrUserDeleted = xerrors.New("user deleted")
```
This error is never thrown outside dbmem right? We don't need to export it outside the package.
```suggestion
var errUserDeleted = xerrors.New("user deleted")
```

hugodutka replied:
Good catch.

Emyrk commented on coderd/database/modelmethods.go:
```diff
@@ -173,7 +174,17 @@ func (v TemplateVersion) RBACObjectNoTemplate() rbac.Object {
+		})
+}
+
+func (gm GroupMember) RBACObject() rbac.Object {
+	return rbac.ResourceGroupMember.WithID(gm.UserID).InOrg(gm.OrganizationID).WithOwner(gm.UserID.String())
```
:+1:

Emyrk commented on coderd/rbac/roles_test.go:
```diff
@@ -112,6 +112,7 @@ func TestRolePermissions(t *testing.T) {
 	// Subjects to user
 	memberMe := authSubject{Name: "member_me", Actor: rbac.Subject{ID: currentUser.String(), Roles: rbac.RoleIdentifiers{rbac.RoleMember()}}}
 	orgMemberMe := authSubject{Name: "org_member_me", Actor: rbac.Subject{ID: currentUser.String(), Roles: rbac.RoleIdentifiers{rbac.RoleMember(), rbac.ScopedRoleOrgMember(orgID)}}}
+	groupMemberMe := authSubject{Name: "group_member_me", Actor: rbac.Subject{ID: currentUser.String(), Roles: rbac.RoleIdentifiers{rbac.RoleMember(), rbac.ScopedRoleOrgMember(orgID)}, Groups: []string{groupID.String()}}}
```
This is great :+1:

Emyrk reviewed (commented):
@hugodutka this is great :+1: 

> I updated groupsauth_test.go to work with the new permission system. I saw that you assumed that if a user didn't have GroupMember read permission then GetGroupMembersByGroupID would return an error, but my implementation returns an empty list instead. I updated the test to match this behavior. Hope that's ok.

:+1: That is fine

-----

My only blocking request is the `dbgen` change mentioned in one of my comments. If you change that, the rest LGTM and I'll merge it in. Appreciate the contribution :partying_face: .

**Just leaving notes for myself**, I'll open some issues for some follow ups after this:

- Reduce the number of fields returned in [`group_members_expanded`](https://github.com/coder/coder/blob/5373dd7d69f7d96f4aa18f6487389ea7fe86cd58/coderd/database/migrations/000242_group_members_view.up.sql#L11-L30). Requires plumbing through the apis and UI. An excess number of fields is currently returned.
  - Ideally we can replace `codersdk.ReducedUser` in `codersdk.Group` with this `database.GroupMember` type?
 - `/groups` api endpoint exists in `/{organization}`. Given https://github.com/coder/coder/issues/14213, we should reconsider a single `/groups` endpoint with filters, or add additional endpoints for fetching groups.

hugodutka committed:
make ErrUserDeleted private

hugodutka committed:
`dbgen.GroupMember` now fails when a supplied user or group doesn't exist

Emyrk reviewed (approved):
:+1: Approved! I'll merge it in. The weekly-docs CI job is failing because docker changed a link :cry:
